File: translate.R
chunk_size = 1000

library(magrittr)
google_get_supported_languages <- function() {
  
  url <- "https://cloud.google.com/translate/docs/languages"
  
  if (!RCurl::url.exists(url)) {
    
    stop("The URL is not available.")
    
  }
  
  
  webpage <- rvest::read_html(url)
  
  table <- webpage %>% rvest::html_nodes('table')
  df <- table %>% rvest::html_table()
  df_lang <- df[[1]] %>% tibble::as_tibble()
  return(df_lang)
  
}

google_is_valid_language_code <- function(language_code) {
  
  
  if (language_code == "auto") {
    return(TRUE)
  }
  
  if (language_code %in% c(google_get_supported_languages()$`ISO-639 code`,"zh-CN","zh-TW")) {
    return(TRUE)
  } else {
    return(FALSE)
  }
}

google_translate_long_text <- function(text, target_language = "zh-CN", source_language = "auto", chunk_size = 1000) {
  if (!google_is_valid_language_code(target_language)) {
    stop("Invalid target language code.")
  }
  if (!google_is_valid_language_code(source_language)) {
    stop("Invalid source language code.")
  }
  
  # Function to split text into chunks
  split_text <- function(text, chunk_size) {
    split_indices <- seq(1, nchar(text), by = chunk_size)
    sapply(split_indices, function(i) substr(text, i, i + chunk_size - 1))
  }
  
  # Split text into chunks if it's too long
  if (nchar(text) > chunk_size) {
    text_chunks <- split_text(text, chunk_size)
  } else {
    text_chunks <- list(text)
  }
  
  # Translate each chunk
  translations <- sapply(text_chunks, function(chunk) {
    formatted_text <- urltools::url_encode(chunk)
    formatted_link <- paste0(
      "https://translate.google.com/m?tl=",
      target_language, "&sl=", source_language,
      "&q=", formatted_text
    )
    
    response <- httr::GET(formatted_link)
    translation <- httr::content(response) %>%
      rvest::html_nodes("div.result-container") %>%
      rvest::html_text()
    
    translation <- urltools::url_decode(translation)
    gsub("\n", "", translation)
  })
  
  # Combine translated chunks
  paste(translations, collapse = " ")
}

File: translate.py
import requests
from bs4 import BeautifulSoup
from urllib.parse import quote_plus, unquote

# Function to get supported languages from Google Translate page
def google_get_supported_languages():
    url = "https://cloud.google.com/translate/docs/languages"
    response = requests.get(url)
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Find the table with the supported languages
    table = soup.find_all('table')[0]
    rows = table.find_all('tr')
    
    # Extract the languages and ISO codes into a dictionary
    languages = {}
    for row in rows[1:]:  # Skip the header row
        cols = row.find_all('td')
        if len(cols) >= 2:
            language_name = cols[0].get_text(strip=True)
            iso_code = cols[1].get_text(strip=True)
            languages[iso_code] = language_name
    
    return languages

# Function to check if a language code is valid
def google_is_valid_language_code(language_code):
    supported_languages = google_get_supported_languages()
    return language_code == "auto" or language_code in supported_languages or language_code in ["zh-CN", "zh-TW"]

# Function to split text into chunks
def split_text(text, chunk_size):
    return [text[i:i+chunk_size] for i in range(0, len(text), chunk_size)]

# Function to translate long text
def google_translate_long_text(text, target_language="zh-CN", source_language="auto", chunk_size=1000):
    if not google_is_valid_language_code(target_language):
        raise ValueError("Invalid target language code.")
    
    if not google_is_valid_language_code(source_language):
        raise ValueError("Invalid source language code.")
    
    # Split the text into chunks if it's too long
    if len(text) > chunk_size:
        text_chunks = split_text(text, chunk_size)
    else:
        text_chunks = [text]
    
    translations = []
    
    # Translate each chunk
    for chunk in text_chunks:
        formatted_text = quote_plus(chunk)  # URL encode the text
        formatted_link = f"https://translate.google.com/m?tl={target_language}&sl={source_language}&q={formatted_text}"
        
        response = requests.get(formatted_link)
        
        if response.status_code == 200:
            # Parse the translated text from the response
            soup = BeautifulSoup(response.text, 'html.parser')
            translation = soup.find('div', class_='result-container').get_text(strip=True)
            translations.append(unquote(translation))  # URL decode the translation
        else:
            raise ValueError(f"Translation failed for chunk: {chunk}")
    
    # Combine translated chunks and return the result
    return " ".join(translations)

